File: linkList/6.sortList/py/sortList.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    start:Optional[ListNode] = None
    end:Optional[ListNode] = None
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        l = 0 
        node = head
        while node:
            l+=1
            node = node.next
        i = 1
        next,lastTeamEnd = None,None
        
        while i < l :
            l1 = head
            r1 = self.findEnd(l1,i)
            l2 = r1.next
            r2 = self.findEnd(l2,i)
            next = r2.next
            r1.next = None
            r2.next = None
            self.merge(l1,r1,l2,r2)
            head = self.start
            lastTeamEnd = self.end
            while next:
                l1 = next
                r1 = self.findEnd(l1,i)
                l2 = r1.next
                if not l2:
                    lastTeamEnd.next = l1
                    break
                r2 = self.findEnd(l2,i)
                next = r2.next
                r1.next = None
                r2.next = None
                self.merge(l1,r1,l2,r2)
                lastTeamEnd.next = self.start
                lastTeamEnd = self.end
            i <<= 1
        return head
            
    def findEnd(self,start:Optional[ListNode],k:int):
        while start.next and k > 1:
            start = start.next
            k -= 1
        return start

    def merge(self,l1:Optional[ListNode],r1:Optional[ListNode],l2:Optional[ListNode],r2:Optional[ListNode]):
        start,end = None,None
        if l1.val < l2.val:
            start = l1
            l1 = l1.next
        else:
            start = l2
            l2 = l2.next
        node = start
        while l1 and l2 :
           if l1.val < l2.val:
               node.next = l1
               l1 = l1.next
               node = node.next
           else:
               node.next = l2
               l2 = l2.next
               node = node.next
        if l1:
            node.next = l1
            end = r1
        else:
            node.next = l2
            end = r2
        self.start = start
        self.end = end

File: linkList/6.sortList/go/sortList.go
package _go

type ListNode struct {
	Val  int
	Next *ListNode
}

// 常数及空间复杂度，不能使用递归
func sortList(head *ListNode) *ListNode {
	// 获取整个链表的长度
	node := head
	l := 0
	for node != nil {
		l++
		node = node.Next
	}
	var l1, r1, l2, r2, start, end, lastTeamEnd, next *ListNode
	// 步长每次乘以二
	for i := 1; i < l; i <<= 1 {
		l1 = head
		r1 = findEnd(l1, i)
		// 因为有记录长度，这里l2不会为空，不用判断
		l2 = r1.Next
		r2 = findEnd(l2, i)
		// 记录下一组
		next = r2.Next
		r1.Next = nil
		r2.Next = nil
		head, lastTeamEnd = merage(l1, r1, l2, r2)
		for next != nil {
			l1 = next
			r1 = findEnd(l1, i)
			l2 = r1.Next
			if l2 == nil {
				lastTeamEnd.Next = l1
				break
			}
			r2 = findEnd(l2, i)
			next = r2.Next
			r1.Next = nil
			r2.Next = nil
			start, end = merage(l1, r1, l2, r2)
			lastTeamEnd.Next = start
			lastTeamEnd = end
		}

	}
	return head

}

// 寻找指定长度的链表区域
func findEnd(start *ListNode, k int) *ListNode {
	for start.Next != nil && k > 1 {
		start = start.Next
		k--
	}
	return start
}

// 归并(整合后返回头和尾)
func merage(l1, r1, l2, r2 *ListNode) (start, end *ListNode) {
	if l1.Val < l2.Val {
		start = l1
		l1 = l1.Next
	} else {
		start = l2
		l2 = l2.Next
	}
	node := start
	for l1 != nil && l2 != nil {
		if l1.Val < l2.Val {
			node.Next = l1
			l1 = l1.Next
			node = node.Next
		} else {
			node.Next = l2
			l2 = l2.Next
			node = node.Next
		}
	}

	if l1 == nil {
		node.Next = l2
		end = r2
	} else {
		node.Next = l1
		end = r1
	}
	return

}
